app/components/message_attachments/action_menu/action_menu.js

```diff
@@ -0,0 +1,174 @@
+import {preventDoubleTap} from 'app/utils/tap';
+import {makeStyleSheetFromTheme, changeOpacity} from 'app/utils/theme';
+
+export default class ActionMenu extends PureComponent {
+    static propTypes = {
+        actions: PropTypes.shape({
+            doPostAction: PropTypes.func.isRequired,
+            setMenuActionSelector: PropTypes.func.isRequired,
+        }).isRequired,
+        id: PropTypes.string.isRequired,
+        name: PropTypes.string.isRequired,
+        dataSource: PropTypes.string,
+        options: PropTypes.arrayOf(PropTypes.object),
+        postId: PropTypes.string.isRequired,
+        theme: PropTypes.object.isRequired,
+        navigator: PropTypes.object,
+    };
+
+    static contextTypes = {
+        intl: intlShape,
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            selectedText: null,
+        };
+    }
+
+    handleSelect = (selected) => {
+        if (!selected) {
+            return;
+        }
+
+        const {dataSource, actions, postId, id} = this.props;
+
+        let selectedText;
+        let selectedValue;
+        if (dataSource === 'users') {
+            selectedText = selected.username;
+            selectedValue = selected.id;
+        } else if (dataSource === 'channels') {
+            selectedText = selected.display_name;
+            selectedValue = selected.id;
+        } else {
+            selectedText = selected.text;
+            selectedValue = selected.value;
+        }
+
+        this.setState({selectedText});
+
+        actions.doPostAction(postId, id, selectedValue);
+    }
+
+    goToMenuActionSelector = preventDoubleTap(() => {
+        const {intl} = this.context;
+        const {navigator, theme, actions, dataSource, options} = this.props;
+
+        actions.setMenuActionSelector(dataSource, this.handleSelect, options);
+
+        navigator.push({
+            backButtonTitle: '',
+            screen: 'MenuActionSelector',
+            title: intl.formatMessage({id: 'action_menu.select', defaultMessage: 'Select an option'}),
+            animated: true,
+            navigatorStyle: {
+                navBarTextColor: theme.sidebarHeaderTextColor,
+                navBarBackgroundColor: theme.sidebarHeaderBg,
+                navBarButtonColor: theme.sidebarHeaderTextColor,
+                screenBackgroundColor: theme.centerChannelBg,
+            },
+        });
+    });
+
+    render() {
+        const {intl} = this.context;
+        const {name, theme, id} = this.props;
+        const {selectedText} = this.state;
+        const style = getStyleSheet(theme);
+
+        let text = name || intl.formatMessage({id: 'action_menu.select', defaultMessage: 'Select an option'});
+        let selectedStyle = style.dropdownPlaceholder;
+        let submitted;
+        if (selectedText) {
+            text = selectedText;
+            selectedStyle = style.dropdownSelected;
+            submitted = [
```

Comment:
You could use a `React.Fragment` here, and I think that would save having to add keys and such